WcaOnRails/app/webpacker/components/Panel/Board/CouncilLeaders.jsx

```diff
@@ -0,0 +1,74 @@
+import React, { useMemo, useState } from 'react';
+import { Icon, Modal, Table } from 'semantic-ui-react';
+import useLoadedData from '../../../lib/hooks/useLoadedData';
+import { rolesOfGroupType, fetchUserGroupsUrl } from '../../../lib/requests/routes.js.erb';
+import { COUNCILS_STATUS, GROUP_TYPE } from '../../../lib/helpers/groups-and-roles-constants';
+import Loading from '../../Requests/Loading';
+import Errored from '../../Requests/Errored';
+import LeaderChangeForm from './LeaderChangeForm';
+
+export default function CouncilLeaders() {
+  const [editCouncil, setEditCouncil] = useState();
+  const councilsFetch = useLoadedData(fetchUserGroupsUrl(GROUP_TYPE.COUNCILS));
+  const councilLeadersFetch = useLoadedData(
+    rolesOfGroupType(GROUP_TYPE.COUNCILS, COUNCILS_STATUS.LEADER),
+  );
+
+  const leaders = useMemo(() => {
+    const leaderMap = {};
+    councilLeadersFetch?.data?.forEach((leader) => {
+      leaderMap[leader.group.id] = leader;
+    });
+    return leaderMap;
+  }, [councilLeadersFetch.data]);
```

Comment:
Rather than create this mapping of group ids to leaders, could we just do a `find` when we need to get a leader from a group id? Something like this (may handle edge cases incorrectly):
```
fun getLeaderOf(council) {
  return councilLeadersFetch?.data?.find((leader) => leader.group.id = council.id
}
```
Or maybe that wouldn't be any better? I might be overthinking this one.

Reply:
The reason why I created map was to reduce time complexity. In your case, if we need to get the leaders of n councils, the complexity will be O(n^2).

But the value of n will be maximum 50 I guess (currently if teams/committees are also considered, it will come to around only 12, expecting big increase in future, it can go up to only 50 I guess).

I've heard that in react or rather in frontend, we don't look much into complexity of smaller ones. I'm bit confused, what is your view in the tradeoffs between clean code and time complexity for frontend?

Reply:
I was thinking about it from a clarity perspective, since I was a bit confused about what was going on when I first read this (partly because I wasn't familiar with the format of the underlying data). Since the object itself was only needed in order to get leaders of councils, it seemed clearer to have that as a function.

But looking at it again, I don't think there's a huge difference in clarity. Especially since we need to get the leader of every group for the table (I was originally only thinking about getting the leader for the editing modal). So I'd say it's fine to leave it alone and ignore my original comment.

(As far as efficiency goes, I don't think there will be any significant difference for an array/object with a few hundred items.)